atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/DefaultSweepStateCoordinatorTest.java

```diff
@@ -0,0 +1,321 @@
+
+        runTaskWithBucket(chosenBucket -> {
+            // The "tail" bucket will be in the non-head set
+            assertThat(chosenBucket).isEqualTo(tail);
+
+            // Now, the "tail" bucket is moved to the head set.
+            buckets.update(Set.of(tail));
+        });
+
+        // We should have removed the "tail" bucket from the candidate set, even if it's moved to the headset.
+        assertThat(coordinator.tryRunTaskWithBucket(chosenBucket -> {})).isEqualTo(SweepOutcome.NOTHING_TO_SWEEP);
+    }
+
+    @Test
+    public void returnsNothingAvailableIfNoBucketUnlocked() {
+        SweepableBucket head = bucket(0, 0);
+        Set<SweepableBucket> sweepableBuckets = Set.of(head);
+        buckets.update(sweepableBuckets);
+        runTaskWithBucket(chosenBucket -> {
+            // chosenBucket (the only bucket) is locked.
+            assertThat(coordinator.tryRunTaskWithBucket(newBucket -> {
+                        throw new RuntimeException("Should not have been called");
+                    }))
+                    .isEqualTo(SweepOutcome.NOTHING_AVAILABLE);
+        });
+    }
+
+    @Test
+    public void returnsNothingToSweepIfNoBucketsInCandidateSet() {
+        assertThat(coordinator.tryRunTaskWithBucket(chosenBucket -> {})).isEqualTo(SweepOutcome.NOTHING_TO_SWEEP);
+    }
+
+    @Test
+    public void requestsRefreshIfNoBucketsRemaining() {
+        coordinator.tryRunTaskWithBucket(chosenBucket -> {});
+        assertThat(retriever.getUpdateRequests()).isEqualTo(1);
+    }
+
+    @Test
+    public void afterHeadBucketsLockedSelectsBucketBasedOnProvidedIterationOrderUntilFindsFirstUnlocked() {
+        List<SweepableBucket> sweepableBuckets =
+                IntStream.range(1, 10).mapToObj(i -> bucket(0, i)).collect(Collectors.toList());
+        SweepableBucket headBucket =
+                bucket(0, 0); // This needs to go at the front since we're always going to visit the head of the sweep
+        Collections.shuffle(sweepableBuckets);
+
+        List<SweepableBucket> rawIterationOrder =
+                Streams.concat(Stream.of(headBucket), sweepableBuckets.stream()).collect(Collectors.toList());
+
+        List<Lockable<SweepableBucket>> iterationOrder = rawIterationOrder.stream()
+                .map(lockableFactory::createLockable)
+                .map(Mockito::spy) // So that we can verify that we only try to lock the first few elements
+                .collect(Collectors.toList());
+        iterationOrderGenerator.set(_buckets -> iterationOrder.stream());
+
+        int numberOfElementsToLockInAdvance = 5;
+        for (int i = 0; i < numberOfElementsToLockInAdvance; i++) {
+            Lockable<SweepableBucket> lockable = iterationOrder.get(i);
+
+            // We don't care about the result of the lock, just that we're locking it.
+            // but CheckReturnValue forces us to make that explicit.
+            Optional<LockedItem<SweepableBucket>> _unused = lockable.tryLock(_ignored -> {});
```

Comment:
nit: usually for these kind of situations I prefer a `@SuppressWarnings` annotation as that communicates intention more clearly / doesn't need some extra code. I get that the scope is different here (in that I think that's a method level thing), though I think I still prefer the other way.